kotlin/src/com/google/idea/blaze/kotlin/sync/importer/BlazeKotlinWorkspaceImporter.java

```diff
@@ -0,0 +1,98 @@
+import com.google.idea.blaze.base.sync.projectview.ProjectViewTargetImportFilter;
+import com.google.idea.blaze.base.targetmaps.TransitiveDependencyMap;
+import com.google.idea.blaze.java.sync.model.BlazeJarLibrary;
+import com.google.idea.blaze.kotlin.sync.model.BlazeKotlinImportResult;
+import com.intellij.openapi.project.Project;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+public class BlazeKotlinWorkspaceImporter {
+    private final TargetMap targetMap;
+    private final ProjectViewTargetImportFilter importFilter;
+
+    public BlazeKotlinWorkspaceImporter(
+            Project project,
+            WorkspaceRoot workspaceRoot,
+            ProjectViewSet projectViewSet,
+            TargetMap targetMap) {
+        this.targetMap = targetMap;
+        importFilter = new ProjectViewTargetImportFilter(project, workspaceRoot, projectViewSet);
+    }
+
+    public BlazeKotlinImportResult importWorkspace() {
+        HashMap<LibraryKey, BlazeJarLibrary> libraries = new HashMap<>();
+        HashMap<TargetIdeInfo, ImmutableList<BlazeJarLibrary>> targetLibraryMap = new HashMap<>();
+
+        collectTransitiveLibsFromKotlinSourceTargets(libraries, targetLibraryMap);
+
+        return new BlazeKotlinImportResult(
+                ImmutableList.copyOf(libraries.values()),
+                ImmutableMap.copyOf(targetLibraryMap));
+    }
+
+    private void collectTransitiveLibsFromKotlinSourceTargets(
+            HashMap<LibraryKey, BlazeJarLibrary> libraries,
+            HashMap<TargetIdeInfo, ImmutableList<BlazeJarLibrary>> targetLibraryMap
+    ) {
+        List<BlazeJarLibrary> scratchLibSet = new ArrayList<>();
+
+        targetMap
+                .targets().stream()
+                .filter(target -> target.kind.languageClass.equals(LanguageClass.KOTLIN))
+                .filter(importFilter::isSourceTarget)
+                .forEach(ideInfo -> withTransitiveTargets(ideInfo).forEach(depIdeInfo -> {
+                    scratchLibSet.clear();
+                    depIdeInfo.javaIdeInfo.jars.stream().map(BlazeJarLibrary::new).forEach(depJar -> {
+                        libraries.putIfAbsent(depJar.key, depJar);
+                        scratchLibSet.add(depJar);
+                    });
+                    if (depIdeInfo.kind.languageClass == LanguageClass.KOTLIN) {
+                        targetLibraryMap.put(depIdeInfo, ImmutableList.copyOf(scratchLibSet));
+                    }
+                }));
+    }
+
+    @Nonnull
+    private Stream<TargetIdeInfo> withTransitiveTargets(TargetIdeInfo target) {
+        return Stream.concat(
+                Stream.of(target),
+                TransitiveDependencyMap.getTransitiveDependencies(target.key, targetMap).stream()
+                        .map(depTarget -> {
```

Comment:
```
            .map(targetMap::get)
            .filter(Objects::nonNull)
            .filter(info -> info.javaIdeInfo != null)
            .filter(Objects::nonNull));
```

Reply:
last filter shouldn't be necessary I think. 
```
                 .map(targetMap::get)
                        .filter(Objects::nonNull)
                        .filter(info -> info.javaIdeInfo != null)
        );
```